src/mscorlib/src/System/String.cs

```diff
@@ -150,7 +186,39 @@ public static String Join<T>(String separator, IEnumerable<T> values)
             }
         }
 
+        [ComVisible(false)]
+        public static String Join<T>(Char separator, IEnumerable<T> values) {
+            if (values == null)
+                throw new ArgumentNullException("values");
+            Contract.Ensures(Contract.Result<String>() != null);
+            Contract.EndContractBlock();
+
+            using(IEnumerator<T> en = values.GetEnumerator()) {
+                if (!en.MoveNext())
+                    return String.Empty;
 
+                StringBuilder result = StringBuilderCache.Acquire();
+                if (en.Current != null) {
+                    // handle the case that the enumeration has null entries
+                    // and the case where their ToString() override is broken
+                    string value = en.Current.ToString();
```

Comment:
Double interface invocation: You are calling `en.Current` twice (in your defense, I think you may have made this PR before #4384 was merged and just copied everything over). It could instead be cached into a local like

```cs
T currentValue = en.Current;
if (currentValue != null)
{
    result.Append(currentValue.ToString());
}
```